### PR TITLE
Introduce civi.api4.authorizeRecord and civi.api4.validate

### PR DESCRIPTION
**Pull request by totten**
Overview
----------------------------------------
Combines prior PRs. Refactor so that events use more consistent style (ie `civi.api4.{$TASK}` with alias `civi.api4.{$TASK}::{$ENTITY}`). Description TODO.

_A brief description of the pull request. Keep technical jargon to a minimum. Hyperlink relevant discussions._

Before
----------------------------------------
_What is the old user-interface or technical-contract (as appropriate)?_
_For optimal clarity, include a concrete example such as a screenshot, GIF ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)), or code-snippet._

After
----------------------------------------
_What changed? What is new old user-interface or technical-contract?_
_For optimal clarity, include a concrete example such as a screenshot, GIF ([LICEcap](http://www.cockos.com/licecap/), [SilentCast](https://github.com/colinkeenan/silentcast)), or code-snippet._

Technical Details
----------------------------------------
_If the PR involves technical details/changes/considerations which would not be manifest to a casual developer skimming the above sections, please describe the details here._

Comments
----------------------------------------
_Anything else you would like the reviewer to note_

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
188 test fails eg
api\v4\Action\BasicActionsTest::testCrud
Civi\API\Exception\UnauthorizedException: ACL check failed

**Comment by totten** [comment]
@eileenmcnaughton Yeah, I think that failure (which recurred in several places) was due to the (lack of) fallback behavior in scenarios where there was no BAO class. Several other failures were due to a copy-paste-typo that I made a few minutes before pushing up. It's doing a new run now. 🤞

**Comment by eileenmcnaughton** [comment]
hmm closer - but still 3 fails - this test should be handled in https://github.com/civicrm/civicrm-core/blob/f4a72e8da90c146109043b27f0ca5f947dfe0519/ext/financialacls/financialacls.php#L157

but it seems the code has not yet been moved to the extension & is sitting here https://github.com/civicrm/civicrm-core/blob/5f31e5f4e02c888944d798be0827bfdb7f096710/api/v3/Contribution.php#L210

api_v3_FinancialTypeACLTest::testDeleteACLContribution
api call should have failed but it succeeded We expected a failure for Contribution delete but got a success: Array
(
    [is_error] => 0
    [version] => 3
    [count] => 1
    [id] => 1
    [values] => Array
        (
            [1] => 1
        )

)

Failed asserting that 0 matches expected 1.

/home/jenkins/bknix-dfl/build/core-20533-3e0yp/web/sites/all/modules/civicrm/Civi/Test/Api3TestTrait.php:86
/home/jenkins/bknix-dfl/build/core-20533-3e0yp/web/sites/all/modules/civicrm/Civi/Test/Api3TestTrait.php:137
/home/jenkins/bknix-dfl/build/core-20533-3e0yp/web/sites/all/modules/civicrm/tests/phpunit/api/v3/FinancialTypeACLTest.php:313
/home/jenkins/bknix-dfl/build/core-20533-3e0yp/web/sites/all/modules/civicrm/tests/phpunit/CiviTest/CiviUnitTestCase.php:237
/home/jenkins/bknix-dfl/extern/phpunit8/phpunit8.phar:671

**Comment by totten** [comment]
> api_v3_FinancialTypeACLTest::testDeleteACLContribution

Yeah, Coleman had already put code for this in the extension. I changed it from a hook-listener to Symfony-listener, but the way I registered it was a little flaky. (It would only work during a solo test-run - not during a full test run.) Switching the registration mechanism has fixed this locally.

**Comment by eileenmcnaughton** [comment]
@totten so in terms of moving that code out of v3 api into the extension nothing is changed by this ?

**Comment by totten** [comment]
@eileenmcnaughton The incantations are changed, but the structure is the same as Coleman's prior PR. So `civicrm_api3_contribution_delete()`  delegates [over here]https://github.com/civicrm/civicrm-core/pull/20533/commits/a98791f0961cdf31511c6ebe6cb5a888b268613e#diff-e615c51bff0c1d7b67d60ac46cfe03e84261c8be20aff18907529796c25b48adR207).

```php
   if (!empty($params['check_permissions']) && !\Civi\Api4\Utils\CoreUtil::checkAccessDelegated('Contribution', 'delete', ['id' => $contributionID], CRM_Core_Session::getLoggedInContactID())) {
```

I suppose one might call it a quirk that an APIv3 implementation calls a utility from APIv4-land, though I don't think anything is lost in doing so. If it seems problematic for some reason, then we just need to rename  `\Civi\Api4\Utils\CoreUtil::checkAccessDelegated()` and tweak to v3/v4. We would also want to rename the event from `civi.api4.authorizeRecord` to `civi.api.authorizeRecord` (*with a footnote that it has very limited applicability to v3*).

**Comment by eileenmcnaughton** [comment]
@totten yeah - I guess my feeling is that line really shouldn't be in the v3 api - ie financialacls are supposed to being transformed into 'just another hook-implemented acl' - the change you cite makes it MORE generic - but it's still tied to contribution.api as opposed to 'any entity might do this'

(This discussion is non-blocking on the main effort)

**Comment by totten** [comment]
jenkins test this please

**Comment by seamuslee001** [comment]
jenkins test this please

**Comment by totten** [comment]
@coleman @eileen I wanna verbalize another thing I encountered.

1. There are two categories of permission checks in play:
    * The coarse-grain/guardian permission (e.g. `civi.api.authorize`, `PermissionCheckSubscriber`, `AbstractAction::isAuthorized()`)
    * The fine-grain/record permission (e.g. `CoreUtil::checkAccess...`, `$bao::_checkAccess`, `civi.api4.authorizeRecord`).
2. There are various `_checkAccess()` functions for specific entities/BAOs -- these are important units for fine-grain/record perms. Nearly all of these do some kind of delegation. (Ex: "To check perms for `Phone.update`, find the `contact_id` and then check perms for `Contact.update`.") 
    * In the prior revision, this delegation step looked like `$granted = $bao::checkAccess($action, ...)`.
    * In the current revision, this delegation looks like `$granted = checkAccessDelegated($entity, $action...)`.
3. There is a disagreement between the coarse/guardian and fine/record mechanisms with respect to `$userID`. For coarse/guardian, the active user is implicitly `CRM_Core_Session::getLoggedInContactID()`. For fine/record mechanisms, the `$userID` is passed as an extra argument.
    * I couldn't see a specific reason/use-case for passing through `$userID`. For all I can tell, `$userID` is always strictly the same as `getLoggedInContactID()`.
    * However, in a general sense, there are times when you want to check permissions on behalf of someone who isn't logged in. And evidently this has real cases in Civi - at least we've done the work to make `CRM_Core_Permission::checkPermission(...)` support alternate users...
    * So even if I couldn't see the case where it was used, I found it plausible to assume it was needed.
4. Sometimes, when doing a fine/record check, it also does a coarse/guardian check.
    * The prior PR and current PR do this a bit differently.
    * There are three cases to consider:
        * Straight-up APIv4 call: if you make a fresh call for `Contact:get()`, it needs to do a check on the self-same call.
        * Side-ways APIv4 call: if you make a fresh call for `Contact:checkAccess()->setAction('get')->execute()`, then it does a check on a slightly different call (same-entity, different-action).
        * Internally delegated call: if you make a call for `Phone::update()`, then it does a delegation to see if you have permission to `Contact::update()` (different-entity, different-action).
    * For each case, the question is: If somebody requests fine/record-level check... *should* it also do a coarse/guardian-check?

How all this mess ties together:

* If the answer is "yes, fine-checks should imply coarse-checks", then `$userID` needs to propagate consistently. The current PR does this... and (non-critically) it should maybe do it a bit better. 
* If the answer is "no, fine-checks should not generally imply coarse-checks", then maybe I've gone a bit overboard by changing the signature of `AbstractAction::isAuthorized()` to `AbstractAction::isAuthorized(?int $userID)`.

**Comment by eileenmcnaughton** [comment]
@totten 
a) my head hurts
b) I think accepting userID is a bit of a missing feature so great to add it
c) I can't quite parse examples of fine checks / coarse checks - do you mean ACLS for the fine checks - eg.
user does not have 'access events' but user has an acl to access event 3 in which case user can access event 3 but not event 2

**Comment by eileenmcnaughton** [comment]
On acls - I think the financial permissions acls work the opposite way to the rest - ie they reduce other-wise given permissions - ug

**Comment by totten** [comment]
a) No kidding :face_with_head_bandage: 

b) My only fear is that it's not so simple -- more pre-existing/unarticulated issues may intervene and prompt us to pivot to a different approach. (An alternative approach - which has been sketched by Coleman before - is to temporarily override the global contact ID - and then set it back.)

c) Right. It's that interplay between things like (i) `access CiviEvent`, `access CiviMail`, `edit all contacts`, etc  (aka coarse-grained/guardian/`civi.api.authorize`/`PermissionCheckSubscriber`/`AbstractAction::isAuthorized()`) and (ii) ACLs, financial ACLs, etc (aka fine-grained/record-level/`civi.api4.authorizeRecord`/`CoreUtil::checkAccessRecord()`).

> On acls - I think the financial permissions acls work the opposite way to the rest - ie they reduce other-wise given permissions - ug

Well... it passes any existing tests... FWIW, this revision can distinguish between 'explicitly authorized' (`$authorized===TRUE`), 'explicitly prohibited' (`$authorized===FALSE`), and 'indeterminate' (`$authorized===NULL`). So for `financialacls`, it does a prohibition via (`$e->setAuthorized(FALSE);`).

--

I was curious to see: *When, if ever, does the system currently do permission-checks on behalf of other users (who are not presently interacting with the app)?* Such use-cases ought to use `CRM_Core_Permission::check($perm, $userID)`, and grepping for that in core+universe (`Permission::check(.*,`) found a few candidates. These cases seemed fairly legit:

* _Authx (ext)_: You can grant a permission like `authenticate via password` or `authenticate via api key`. At the moment when a user submits their password, they have not logged in *yet*, but you still need to check the corresponding permission.
* _ACL Cache/API_: The signatures for `CRM_ACL_API::whereClause()` and `CRM_Contact_BAO_Contact_Permission::cache()` allow you to populate the ACL cache on behalf of a non-present user. I couldn't see a code-path which takes advantage of this today. (*They all seemed to lean on `getLoggedInContactID()`.*), but maybe I just missed it. Even if it's not there today, it could make sense someday. Ex: If you implemented an ACL-cache based on background-updates, then it could loop through a bunch of non-present staff-users and try to update/recompute their ACLs.
* _Group Admin (ext)_: This hooks into ACLs. Since the ACL Cache/API allows lookups for non-present users, it has to follow suit.
* _Afform (ext)_: This hooks into `Permission::check()`. Since `Permission::check()` allows lookups for non-present users, it has to follow suit.

Maybe the ACL cache example is salient? Suppose you had a background-script proactively populating the ACL cache for staff-users - and the ACLs depended on smart-groups. The smart-groups are built from APIv4 queries. It would be important for all the permissions to be resolved. OTOH, for ACL cache, you'd probably want to just do the `get()` action as the target-user... 

--

Speaking of ACLs - there are ACLs for CustomGroups. Is `CRM_Core_BAO_CustomValue::_checkAccess()` doing the right thing by delegating the access-check to `Contact.update`? Shouldn't it actually be consulting the CustomGroup ACLs?

**Comment by eileenmcnaughton** [comment]
Re customGroups - I think the acls have traditionally done this really weirdly - ie our favourity function 'getTree' sets permissions to true (often inappropriately) and returns metadata without acl-limited custom fields - I suspect that is behind some of the test fails in this PR https://github.com/civicrm/civicrm-core/pull/20531

My head still hurts - but I'm not sure if any of the edge cases invalidate the approach here

**Comment by colemanw** [comment]
Eileen raises a good point about gatekeeper vs fine-grained checks - for most entities the latter overrides the former, which means we should *not* be doing gatekeeer checks in the api at all for any entity which supports acls.
This is already the case for Contacts, where api permissions are set to `[]` (empty array implying no permission needed).
My intention was to do the same for other entities one-by-one as we get acl checks working for them in the api, and I was hoping this pr would take care of a lot of them.

**Comment by totten** [comment]
@colemanw I've pushed up both changes. I'll work on the description tomorrow - but it should be testable now.

File: CRM/Core/BAO/CustomValue.php [hunk 1]
```diff
@@ -221,4 +221,59 @@ public function addSelectWhereClause() {
...
+    if (!$eid) {
+      $tableName = CRM_Core_DAO::getFieldValue('CRM_Core_DAO_CustomGroup', $groupName, 'table_name', 'name');
+      $eid = CRM_Core_DAO::singleValueQuery("SELECT entity_id FROM `$tableName` WHERE id = " . (int) $record['id']);
+    }
+
+    // Do we have access to the target record?
+    if (in_array($extends, ['Contact', 'Individual', 'Organization', 'Household'])) {
+      return \Civi\Api4\Utils\CoreUtil::checkAccessDelegated('Contact', 'update', ['id' => $eid], $userID);
+    }
+    elseif (class_exists('Civi\\Api4\\' . $extends)) {
```
Comment (colemanw) [inline CRM/Core/BAO/CustomValue.php:269]: What about oddball `CiviCase` class?
I think this ought to be 
```suggestion
    elseif (\Civi\Api4\Utils\CoreUtil::getApiClass($extends)) {
```

**Comment by colemanw** [comment]
This is looking good and is passing tests. Feels safe to merge at this point & we can hammer on it during the RC.

**Comment by eileenmcnaughton** [comment]
@totten it is merged - let the great branching commence

**Pull request merged**